InteractionRegions: improve region consolidation

etiennesegonzac opened the pull request:
#### 9ac8cd985eb6d6e2f366ccd610fbd4d9fc0267b4
<pre>
InteractionRegions: improve region consolidation
<a href="https://bugs.webkit.org/show_bug.cgi?id=258026">https://bugs.webkit.org/show_bug.cgi?id=258026</a>
&lt;rdar://110283650&gt;

Reviewed by Megan Gardner.

Introduce a new InteractionRegion consolidation method aimed at removing
superfluous regions generated for structural elements nested in an
interactive container. Typically, when the container and everything
inside has `cursor: pointer`.
This requires looking at the ancestors (are we contained by an existing
region?) and children (do we discover a nested region in the subtree
later on?).

* Source/WebCore/rendering/EventRegion.h:
Add new HashSets for containers removal tracking.
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::uniteInteractionRegions):
Clean-up, don&apos;t keep track of rects for interactions we consolidate.
(WebCore::EventRegionContext::shouldConsolidateInteractionRegion):
Add potential containers to the candidates set.
Flag containers for removal if we discover a nested interaction in
their subtree.
Add comments explaining the logic.
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
Filter out the regions flagged for removal before doing the shrink wrap.

* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions.html:
Add a new test case for this scenario.

Canonical link: <a href="https://commits.webkit.org/265186@main">https://commits.webkit.org/265186@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bbf72ce77faf4a773e37353d4835a74fd5220e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10347 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11754 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9787 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10112 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12337 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10299 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12708 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10258 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11026 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8546 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12139 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8331 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9157 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16472 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9432 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9308 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12569 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9789 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7950 "2 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9052 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2439 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13186 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9590 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -535,9 +557,13 @@ bool EventRegion::containsEditableElementsInRect(const IntRect& rect) const
 #if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
 
-void EventRegion::appendInteractionRegions(const Vector<InteractionRegion>& interactionRegions)
+void EventRegion::appendInteractionRegions(const Vector<InteractionRegion>& interactionRegions, const HashSet<ElementIdentifier>& containersToRemove)
```
This feels like a weird shape for this function to take. We should either remove these regions before appending, or change the name to cullThenAppendInteractionRegions or something else like that. This has just taken on a very different shape, and I don't think I like it.

etiennesegonzac replied:
Totally! Plus we should remove them before shrink wrapping anyway.

megangardner reviewed (approved):
I like that better, thanks!

etiennesegonzac committed:
InteractionRegions: improve region consolidation
https://bugs.webkit.org/show_bug.cgi?id=258026
<rdar://110283650>

Reviewed by Megan Gardner.

Introduce a new InteractionRegion consolidation method aimed at removing
superfluous regions generated for structural elements nested in an
interactive container. Typically, when the container and everything
inside has `cursor: pointer`.
This requires looking at the ancestors (are we contained by an existing
region?) and children (do we discover a nested region in the subtree
later on?).

* Source/WebCore/rendering/EventRegion.h:
Add new HashSets for containers removal tracking.
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::uniteInteractionRegions):
Clean-up, don't keep track of rects for interactions we consolidate.
(WebCore::EventRegionContext::shouldConsolidateInteractionRegion):
Add potential containers to the candidates set.
Flag containers for removal if we discover a nested interaction in
their subtree.
Add comments explaining the logic.
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
Filter out the regions flagged for removal before doing the shrink wrap.

* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions.html:
Add a new test case for this scenario.

Canonical link: https://commits.webkit.org/265186@main
